uid/src/main/java/com/tomtom/kotlin/uid/Uid.kt

```diff
@@ -0,0 +1,225 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.tomtom.kotlin.uid
+
+import java.util.*
+
+/**
+ * Generic immutable Unique ID class. Really just an abstraction of UUIDs. Used to unique
+ * identify things. No 2 Uid objects shall be equal unless they are the same instance of the
+ * Uid class or their underlying UUIDs are equal.
+ *
+ * The class is templated to allow creating typesafe Uid's, like Uid<Message> or Uid<Person>.
+ * The class represents UUID as Strings internally, to avoid loads of UUID to String conversions
+ * all the time. This makes the class considerably faster in use than the regular [UUID] class.
+ */
+class Uid<T> {
+    private val uuid: String
+
+    /**
+     * Create a new, unique UUID-based ID.
+     */
+    constructor() {
+        uuid = UUID.randomUUID().toString()
+    }
+
+    /**
+     * Instantiates a Uid with a string. Mainly used when de-serializing existing entities.
+     *
+     * The format of 'uuid' is checked to comply with a standard UUID format, which is:
+     * - Dashes at positions 8, 13, 18, 23 (base 0).
+     * - Characters 0-9 and a-f (lowercase only).
+     *
+     * if this form is used, the creation of the Uid is very fast. If an alternative format
+     * us used, as accepted by [fromString], the call is much more expensive.
+     *
+     * @param uuid Existing string id.
+     * @throws IllegalArgumentException If name does not conform to the string representation
+     * as described in [UUID.toString]. Use [isValid] to make sure the string is valid.
+     */
+    constructor(uuid: String) {
+        /**
+         * This code has been optimized to NOT just call UUID.fromString(uuid) to convert the
+         * UUID-String into a String (and catch an IllegalArgumentException).
+         *
+         * If the UUID does not comply, the expensive call to UUID.fromString is made after all.
+         */
+        val length = uuid.length
+        kotlin.require(
+            UUID_MIN_LENGTH <= length && length <= UUID_MAX_LENGTH
+        ) {
+            "Length of UUID must be [" + UUID_MIN_LENGTH + ", " +
+                UUID_MAX_LENGTH + "], but is " + uuid.length + ", uuid=" + uuid
+        }
+
+        // Check dashes.
+        for (i in UUID_DASH_POS) {
+            if (length < i + 1 || uuid[i] != UUID_DASH) {
+
+                // This will throw an IllegalArgumentException if it went wrong.
+                this.uuid = UUID.fromString(uuid).toString().toLowerCase()
+                return
+            }
+        }
+
+        // Make sure the UUID is lowercase.
+        this.uuid = uuid.toLowerCase()
+
+        // UUID seems to be OK.
+        kotlin.require(onlyContainsValidUUIDCharacters(this.uuid)) {
```

Comment:
Please move check before assignment. The assignment might have unintended side affects when a custom setter is used in the future. Okeej. unlikely, but jah.

Reply:
The UUID needs to be converted to lower-case first. Placing the check first would require that to happen twice (in the check and in the assignment). Uppercase characters are NOT part of a valid Uid.

Reply:
Best of both worlds:
```
        this.uuid = uuid.toLowerCase()
            .also {
                require(onlyContainsValidUUIDCharacters(this.uuid))
            }

Reply:
Done.